### ISSUE TITLE
[iOS] [a11y] Unable to navigate to bottom sliver in CustomScrollView with VoiceOver

### ISSUE THREAD

**Issue by ChristianEdwardPadilla** [flutter/flutter#142065]
Internal: b/321855070

### Steps to reproduce

1. Run the sample code.
2. Enable VoiceOver.
3. Swipe right to focus successive Flutter icon buttons.
4. Continue to swipe right until you reach the end of the list (a larger Flutter icon).

### Expected results

You should be able to reach the larger Flutter icon button by swiping right.

### Actual results

You cannot reach the bottom of the list.

Note: This is iOS-only. I tested Android and web (Chrome on macOS) and you can reach the end of the list with voice assist enabled there.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const SliverFillRemainingExampleApp());

class SliverFillRemainingExampleApp extends StatelessWidget {
  const SliverFillRemainingExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text('SliverFillRemaining Sample')),
        body: const SliverFillRemainingExample(),
      ),
    );
  }
}

class SliverFillRemainingExample extends StatelessWidget {
  const SliverFillRemainingExample({super.key});

  @override
  Widget build(BuildContext context) {
    return CustomScrollView(
      cacheExtent: 9999,
      slivers: [
        SliverList(
          delegate: SliverChildListDelegate([
            const SizedBox(height: 12),
            ElevatedButton(
              onPressed: () {},
              child: FlutterLogo(
                size: 100,
              ),
            ),
            const SizedBox(height: 12),
            ElevatedButton(
              onPressed: () {},
              child: FlutterLogo(
                size: 100,
              ),
            ),
            const SizedBox(height: 8),
            ElevatedButton(
              onPressed: () {},
              child: FlutterLogo(
                size: 100,
              ),
            ),
            const SizedBox(height: 20),
            ElevatedButton(
              onPressed: () {},
              child: FlutterLogo(
                size: 100,
              ),
            ),
            const SizedBox(height: 16),
            ElevatedButton(
              onPressed: () {},
              child: FlutterLogo(
                size: 100,
              ),
            ),
            const SizedBox(height: 16),
            ElevatedButton(
              onPressed: () {},
              child: FlutterLogo(
                size: 100,
              ),
            ),
            const SizedBox(height: 16),
            ElevatedButton(
              onPressed: () {},
              child: FlutterLogo(
                size: 100,
              ),
            ),
            const SizedBox(height: 16),
            ElevatedButton(
              onPressed: () {},
              child: FlutterLogo(
                size: 100,
              ),
            ),
            // Add more buttons to the list if the button below is initially
            // in the viewport.
          ]),
        ),
        SliverFillRemaining(
          hasScrollBody: false,
          child: ElevatedButton(
            onPressed: () {},
            child: FlutterLogo(
              size: 300,
            ),
          ),
        ),
      ],
    );
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel google3, on macOS 14.2.1 23C71 darwin-x64, locale en)
    • Framework revision 1997bec685 (4 days ago), 2024-01-18T00:00:00.000
    • Engine revision b75d6d80d8
    • Dart version d6c08fa9bb

[✓] Android toolchain - develop for Android devices (Android SDK version Stable)
    • Android SDK at google3
    • Platform Stable, build-tools Stable
    • Java binary at: /Applications/Android Studio with
      Blaze.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.85.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.74.0

[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A240d

[✓] Connected device (1 available)
    • Pixel 4 (mobile) • xx • android-arm64 • Android 13 (API 33)
```

</details>

**Comment by jmagman** [comment]
I'm going to drop this down to a P1, "This is a bug on production app which affect a great amount of users" doesn't align with our P0 priority.

**Comment by vashworth** [comment]
So here's what I've figured out so far:

The framework passes a list of semantic nodes and actions to the engine

https://github.com/flutter/engine/blob/abb51c48e26110060d591ed8341bf8bf7987e7ea/lib/ui/platform_dispatcher.dart#L855-L858

And the engine then creates `SemanticObject`/`UIAccessibilityElement` elements that iOS recognizes

https://github.com/flutter/engine/blob/abb51c48e26110060d591ed8341bf8bf7987e7ea/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm#L86

The problem is that the framework is not sending a node for the button in the `SliverFillRemaining`, which I tracked down to this code:

https://github.com/flutter/flutter/blob/f01ce9f4cb41beff7b85122b5fcf1228bb655a87/packages/flutter/lib/src/rendering/viewport.dart#L311-L315
https://github.com/flutter/flutter/blob/4280d29534dad3d035c6b0a6a33db591bd9006bf/packages/flutter/lib/src/rendering/object.dart#L3663

`childrenInPaintOrder` has two slivers, however, the `SliverFillRemaining` one is both not visible and doesn't have a cacheExtent greater than zero, which causes it to get excluded. 

I think perhaps the `cacheExtent` isn't getting calculated properly - needs more investigation.

**Comment by vashworth** [comment]
`cacheExtent` is never set for `SliverFillRemaining` widgets, therefore if it's not visible on start, the semantic nodes for it will not be passed to the engine until it's scrolled into view.

https://github.com/flutter/flutter/blob/4280d29534dad3d035c6b0a6a33db591bd9006bf/packages/flutter/lib/src/rendering/sliver_fill.dart#L168-L173

**Comment by jmagman** [comment]
> `cacheExtent` is never set for `SliverFillRemaining` widgets, therefore if it's not visible on start, the semantic nodes for it will not be passed to the engine until it's scrolled into view.

That matches with what's indicated in b/321855070#comment3

**Comment by vashworth** [comment]
Okay so I have a fix for the issue of `cacheExtent` not being set for `SliverFillRemaining` widgets: https://github.com/flutter/flutter/pull/143612.

However, `RenderSliverFillRemainingWithScrollable` has an addition issue that I'm not sure how to fix.

@Piinks maybe you could advise on this.


So when `RenderSliverFillRemainingWithScrollable` is out of the viewport, the `extent` is calculated to 0, which in turn gives the child a `maxExtent`/`maxHeight` of 0.

https://github.com/flutter/flutter/blob/383067fc93f4ef9867caf027997b3cda9584e2d2/packages/flutter/lib/src/rendering/sliver_fill.dart#L84-L91

Because the `maxHeight` is 0, the `size` is therefore 0 and also the `semanticBounds`.

https://github.com/flutter/flutter/blob/383067fc93f4ef9867caf027997b3cda9584e2d2/packages/flutter/lib/src/rendering/box.dart#L2117

And because the `semanticBounds` are 0, the `_rect` (bounding box) for the semantic node is 0:
https://github.com/flutter/flutter/blob/383067fc93f4ef9867caf027997b3cda9584e2d2/packages/flutter/lib/src/rendering/object.dart#L5123-L5124

Which will cause the node to get dropped from the semantic tree and therefore can't be swiped to:
https://github.com/flutter/flutter/blob/383067fc93f4ef9867caf027997b3cda9584e2d2/packages/flutter/lib/src/rendering/object.dart#L4907-L4909

https://github.com/flutter/flutter/blob/383067fc93f4ef9867caf027997b3cda9584e2d2/packages/flutter/lib/src/rendering/object.dart#L5229-L5234

**Comment by Piinks** [comment]
> So when RenderSliverFillRemainingWithScrollable is out of the viewport, the extent is calculated to 0, which in turn gives the child a maxExtent/maxHeight of 0.

Is the RenderSliverFillRemainingWithScrollable in this scenario still within the cacheExtent? If so, it should be allotted whatever portion of the remainingCacheExtent is available for its extent. If it is outside of the cacheExtent.. I am not sure what size it should be. Maybe we can build in a min extent so that it is always accessible?

**Comment by jmagman** [comment]
If the entire fix is in `packages/flutter/lib/src/rendering/sliver_fill.dart`, any idea why this only reproduced on iOS?

> Note: This is iOS-only. I tested Android and web (Chrome on macOS) and you can reach the end of the list with voice assist enabled there.

**Comment by hangyujin** [comment]
why this only reproduced on iOS >> 
On android a11y mode,  scroll on a sliver list, when scroll to the edge, it looks like android will load the next page of items, iOS just load the next item, maybe that's why?

https://github.com/flutter/flutter/assets/108393416/cb8a0685-df62-4a76-a3dd-8cabdebc9b08

https://github.com/flutter/flutter/issues/143437 looks similar to this one, add a cacheExtent to sliverAppbar will also make the sliverAppbar appear in the semantics tree, but it's still not displayed.

**Comment by vashworth** [comment]
> If the entire fix is in `packages/flutter/lib/src/rendering/sliver_fill.dart`, any idea why this only reproduced on iOS?
> 
> > Note: This is iOS-only. I tested Android and web (Chrome on macOS) and you can reach the end of the list with voice assist enabled there.

tldr; So the accessibility "swipe to select next item" gesture behaves slightly differently between platforms. On Android, if there isn't a next item in the semantic tree, it'll trigger a `ACTION_SCROLL_FORWARD` event, which the engine then uses to send a `SCROLL_UP` event to the framework. The `SCROLL_UP` event will then make the next item visible, which will cause it to get added to the semantic tree.

Whereas on iOS, if there isn't a next item in the semantic tree, no interceptable events happen, so we have no indication to call the framework. I believe part of the purpose of `cacheExtent` is to accommodate for this limitation on iOS as explained in description of https://github.com/flutter/flutter/pull/17021, https://github.com/flutter/flutter/issues/11983#issuecomment-365764023,  https://api.flutter.dev/flutter/rendering/RenderViewportBase/cacheExtent.html.

----

On Android, when you swipe to select the next item, if there is a next item in the semantic tree, it'll trigger a [`ACTION_SHOW_ON_SCREEN` action](https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.AccessibilityAction#ACTION_SHOW_ON_SCREEN).
https://github.com/flutter/engine/blob/1ae2c10e807148737e8220fa5514daa7f32e5502/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L1188-L1192

Which will dispatch the `Action.SHOW_ON_SCREEN` event to the framework:
https://github.com/flutter/flutter/blob/3ae71f5c4334e3f59f722d91c0c8cb6e82cfd20c/packages/flutter/lib/src/rendering/binding.dart#L429-L434

However, if there's not a next item in the semantic tree (as in this issue), it'll trigger a [`ACTION_SCROLL_FORWARD` action](https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo#ACTION_SCROLL_FORWARD).

https://github.com/flutter/engine/blob/1ae2c10e807148737e8220fa5514daa7f32e5502/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L1079-L1096

Which will dispatch the `Action.SCROLL_UP` event to the framework:

https://github.com/flutter/flutter/blob/3538e4c788e106b76649f99758c39ebafa03b8ea/packages/flutter/lib/src/rendering/proxy_box.dart#L4071-L4079

Which will cause the semantic tree to get updated and create the SemanticNode for the `SliverFillRemaining`, since it's now visible.

On iOS, when you swipe to select the next item, if there is a next item in the semantic tree, it'll trigger a `accessibilityScrollToVisibleWithChild` method.

https://github.com/flutter/engine/blob/1ae2c10e807148737e8220fa5514daa7f32e5502/shell/platform/darwin/ios/framework/Source/SemanticsObject.mm#L587-L600

Which, same as Android, will dispatch the `Action.SHOW_ON_SCREEN` event to the framework.

However, if there's not a next item in the semantic tree (as in this issue), iOS does not trigger any events that we can intercept it seems.

I believe part of the purpose of `cacheExtent` is to accommodate for this limitation on iOS as explained in description of https://github.com/flutter/flutter/pull/17021, https://github.com/flutter/flutter/issues/11983#issuecomment-365764023,  https://api.flutter.dev/flutter/rendering/RenderViewportBase/cacheExtent.html.
